File: spark/iterator/src/main/java/org/apache/accumulo/spark/AvroUtil.java
package org.apache.accumulo.spark;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.apache.avro.Schema;
import org.apache.avro.SchemaBuilder;

public class AvroUtil {
	private static SchemaBuilder.FieldAssembler<Schema> addAvroField(SchemaBuilder.FieldAssembler<Schema> builder,
			String type, String name) {
		switch (type.toUpperCase()) {
		case "STRING":
			return builder.optionalString(name);

		case "LONG":
			return builder.optionalLong(name);

		case "INTEGER":
			return builder.optionalInt(name);

		case "DOUBLE":
			return builder.optionalDouble(name);

		case "FLOAT":
			return builder.optionalFloat(name);

		case "BOOLEAN":
			return builder.optionalBoolean(name);

		case "BYTES":
			return builder.optionalBytes(name);

		default:
			throw new IllegalArgumentException("Unsupported type '" + type + "'");
		}
	}

	private static SchemaBuilder.FieldAssembler<Schema> closeFieldAssembler(
			SchemaBuilder.FieldAssembler<Schema> rootAssembler,
			SchemaBuilder.FieldAssembler<Schema> columnFieldsAssembler, String columnFamily) {

		if (columnFieldsAssembler == null)
			return rootAssembler;

		// add nested type to to root assembler
		return rootAssembler.name(columnFamily).type(columnFieldsAssembler.endRecord()).noDefault();
	}

	public static Schema buildSchema(SchemaMappingField[] schemaMappingFields) {
		// construct schema
		SchemaBuilder.FieldAssembler<Schema> rootAssembler = SchemaBuilder.record("root").fields();

		// note that the order needs to be exactly in-sync with the avro schema
		// generated on the MMLSpark/Scala side
		String lastColumnFamily = null;
		SchemaBuilder.FieldAssembler<Schema> columnFieldsAssembler = null;
		for (SchemaMappingField schemaMappingField : schemaMappingFields) {

			String columnFamily = schemaMappingField.getColumnFamily();
			String columnQualifier = schemaMappingField.getColumnQualifier();
			String type = schemaMappingField.getType();

			if (columnQualifier != null) {
				if (lastColumnFamily == null || !lastColumnFamily.equals(columnFamily)) {

					// close previous record
					rootAssembler = closeFieldAssembler(rootAssembler, columnFieldsAssembler, lastColumnFamily);

					// open new record
					columnFieldsAssembler = SchemaBuilder.record(columnFamily).fields();
				}

				// add the current field
				columnFieldsAssembler = addAvroField(columnFieldsAssembler, type, columnQualifier);
			} else {
				// close previous record
				rootAssembler = closeFieldAssembler(rootAssembler, columnFieldsAssembler, lastColumnFamily);
				columnFieldsAssembler = null;

				// add the top-level field
				rootAssembler = addAvroField(rootAssembler, type, columnFamily);
			}

			lastColumnFamily = columnFamily;
		}

		rootAssembler = closeFieldAssembler(rootAssembler, columnFieldsAssembler, lastColumnFamily);

		// setup serialization
		return rootAssembler.endRecord();
	}
}

File: connector/datasource/src/main/scala/com/microsoft/accumulo/AvroUtil.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.microsoft.accumulo

import org.apache.avro.{Schema, SchemaBuilder}
import org.apache.spark.sql.types.{DataType, DataTypes, StructField, StructType}
import org.codehaus.jackson.map.ObjectMapper
import org.codehaus.jackson.map.annotate.JsonSerialize.Inclusion

import scala.beans.BeanProperty

// keeping the property names short to not hit any limits
case class RowBuilderField(@BeanProperty cf: String,  // column family
                           @BeanProperty cq: String,  // column qualifier
                           @BeanProperty fvn: String, // filter variable name
                           @BeanProperty t: String,   // type
                           @BeanProperty o: Boolean   // output
                          )

case class JsonSchema(json: String, attributeToVariableMapping: Map[String, String])

@SerialVersionUID(1L)
object AvroUtil {
  def catalystSchemaToJson(inputSchema: StructType): JsonSchema = catalystSchemaToJson(inputSchema, inputSchema)

  def catalystSchemaToJson(inputSchema: StructType, outputSchema: StructType): JsonSchema = {

    var attributeToVariableMapping = scala.collection.mutable.Map[String,  String]()

    var i = 0
    val selectedFields = inputSchema.fields.flatMap(cf => {
      val outputField = outputSchema.find(f => f.name == cf.name)

      cf.dataType match {
        case cft: StructType => cft.fields.map(cq =>
          RowBuilderField(
            cf.name,
            cq.name,
            {
              val variableName = s"v$i"
              attributeToVariableMapping += (s"${cf.name}.${cq.name}" -> variableName)
              i += 1

              variableName
            },
            // TODO: toUpperCase() is weird...
            cq.dataType.typeName.toUpperCase,
            // either the column family is not need -> output = false
            // otherwise we need to check if the column qualifier is present in the output list
            if (outputField.isEmpty) false else outputField.get.dataType.asInstanceOf[StructType].exists(f => f.name == cq.name)
          )
        )
        case _: DataType => Seq(RowBuilderField(
          cf.name,
          null,
          {
            val variableName = s"v$i"
            attributeToVariableMapping += (s"${cf.name}" -> variableName)
            i += 1

            variableName
          },
          // TODO: toUpperCase() is weird...
          cf.dataType.typeName.toUpperCase,
          outputField.isDefined
        ))
      }
    })

    try {
      val mapper = new ObjectMapper()

      // disable serialization of null-values
      mapper.setSerializationInclusion(Inclusion.NON_NULL)

      JsonSchema(mapper.writeValueAsString(selectedFields), attributeToVariableMapping.toMap)
    } catch {
      case e: Exception =>
        throw new IllegalArgumentException(e)
    }
  }

  implicit class CatalystSchemaToAvroRecordBuilder(builder: SchemaBuilder.FieldAssembler[Schema]) {
    def addAvroRecordField(field: StructField): SchemaBuilder.FieldAssembler[Schema] = {
      (field.dataType, field.nullable) match {
          case (DataTypes.BinaryType, true) => builder.optionalBytes(field.name)
          case (DataTypes.BinaryType, false) => builder.requiredBytes(field.name)
          case (DataTypes.BooleanType, true) => builder.optionalBoolean(field.name)
          case (DataTypes.BooleanType, false) => builder.requiredBoolean(field.name)
          case (DataTypes.DoubleType, true) => builder.optionalDouble(field.name)
          case (DataTypes.DoubleType, false) => builder.requiredDouble(field.name)
          case (DataTypes.FloatType, true) => builder.optionalFloat(field.name)
          case (DataTypes.FloatType, false) => builder.requiredFloat(field.name)
          case (DataTypes.IntegerType, true) => builder.optionalInt(field.name)
          case (DataTypes.IntegerType, false) => builder.requiredInt(field.name)
          case (DataTypes.LongType, true) => builder.optionalLong(field.name)
          case (DataTypes.LongType, false) => builder.requiredLong(field.name)
          case (DataTypes.StringType, true) => builder.optionalString(field.name)
          case (DataTypes.StringType, false) => builder.requiredString(field.name)
          // TODO: date/time support?
          case _ => throw new UnsupportedOperationException(s"Unsupported type: $field.dataType")
      }
    }

    def addAvroRecordFields(schema: StructType): SchemaBuilder.FieldAssembler[Schema] = {
      schema.fields.foldLeft(builder) { (builder, field) => builder.addAvroRecordField(field) }
    }
  }

  def catalystSchemaToAvroSchema(schema: StructType): Schema = {
    val fieldBuilder = SchemaBuilder.record("root")
      .fields()

    schema.fields.foldLeft(fieldBuilder) { (_, field) =>
        field.dataType match {
          // nested fields
          case cft: StructType =>
            fieldBuilder
              .name(field.name)
              .`type`(SchemaBuilder
                .record(field.name)
                .fields
                .addAvroRecordFields(cft)
                .endRecord())
              .noDefault()
          // top level fields
          case _ => fieldBuilder.addAvroRecordField(field)
        }
      }
      .endRecord()
  }
}
